### PR TITLE
WH: Verify event counts in load file gen and export stages

### PR DESCRIPTION
**Pull request by psrikanth88**
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: warehouse/upload.go [hunk 1]
```diff
@@ -1271,22 +1352,51 @@ func (job *UploadJobT) areIdentityTablesLoadFilesGenerated() (generated bool, er
...
+				location,
+				row_number() OVER (PARTITION BY staging_file_id, table_name ORDER BY id DESC) AS row_number
+				FROM %[1]s
+				WHERE %[1]s.id >= %[2]v AND %[1]s.id <= %[3]v AND %[1]s.source_id='%[4]s' AND %[1]s.destination_id='%[5]s'
+		)
+		SELECT location
+			FROM row_numbered_load_files
+			WHERE
+				row_number=1`,
+		warehouseutils.WarehouseLoadFilesTable, startID, endID, job.warehouse.Source.ID, job.warehouse.Destination.ID)
```
Comment (sivashanmukh) [inline warehouse/upload.go:1399]: Merge these two funcs

**Review by sivashanmukh — CHANGES_REQUESTED** [review]

File: warehouse/upload.go [hunk 2]
```diff
@@ -215,6 +216,49 @@ func (job *UploadJobT) syncRemoteSchema() (hasSchemaChanged bool, err error) {
...
+	}
+	return
+}
+
+func (job *UploadJobT) matchRowsInStagingAndLoadFiles() {
+	rowsInStagingFiles := job.getTotalRowsInStagingFiles()
+	rowsInLoadFiles := job.getTotalRowsInLoadFiles()
+	if (rowsInStagingFiles != rowsInLoadFiles) || rowsInStagingFiles == 0 || rowsInLoadFiles == 0 {
+		pkgLogger.Errorf(`Error: Rows count mismatch between staging and load files for upload:%d. rowsInStagingFiles: %d, rowsInLoadFiles: %d`, job.upload.ID, rowsInStagingFiles, rowsInLoadFiles)
+		job.counterStat("warehouse_staging_load_file_events_count_mismatched").Increment()
```
Comment (sivashanmukh) [inline warehouse/upload.go:258]: Instead gauge the difference

File: warehouse/upload.go [hunk 3]
```diff
@@ -667,12 +713,35 @@ func (job *UploadJobT) loadTable(tName string) (alteredSchema bool, err error) {
...
+	if generateTableLoadCountVerificationsMetrics {
+		var countErr error
+		totalAfterLoad, countErr = job.whManager.GetTotalCountInTable(tName)
+		if countErr != nil {
+			pkgLogger.Errorf(`Error getting total count in table:%s after load: %v`, tName, countErr)
+		}
+		job.counterStat(`pre_load_table_rows`, tag{name: "tableName", value: strings.ToLower(tName)}).Count(int(totalBeforeLoad))
+		eventsInTableUpload := tableUpload.getTotalEvents()
+		job.counterStat(`post_load_table_rows_estimate`, tag{name: "tableName", value: strings.ToLower(tName)}).Count(int(totalBeforeLoad + eventsInTableUpload))
+		job.counterStat(`post_load_table_rows`, tag{name: "tableName", value: strings.ToLower(tName)}).Count(int(totalAfterLoad))
```
Comment (sivashanmukh) [inline warehouse/upload.go:742]: Lets use Gauge here as well

**Review by sivashanmukh — CHANGES_REQUESTED** [review]
Guage metrics

File: warehouse/table_upload.go [hunk 4]
```diff
@@ -85,6 +86,13 @@ func (tableUpload *TableUploadT) setStatus(status string) (err error) {
 	return err
 }
 
+func (tableUpload *TableUploadT) getTotalEvents() int64 {
+	sqlStatement := fmt.Sprintf(`SELECT total_events FROM %s WHERE wh_upload_id=%v AND table_name='%s'`, warehouseutils.WarehouseTableUploadsTable, tableUpload.uploadID, tableUpload.tableName)
```
[the remark this answers is not in the document]
Comment (psrikanth88) [reply warehouse/table_upload.go:90]: done

**Review by sivashanmukh — APPROVED** [review]
LGTM

File: warehouse/bigquery/bigquery.go [hunk 5]
```diff
@@ -473,6 +473,22 @@ func (bq *HandleT) DownloadIdentityRules(*misc.GZipWriter) (err error) {
 	return
 }
 
+func (bq *HandleT) GetTotalCountInTable(tableName string) (total int64, err error) {
+	sqlStatement := fmt.Sprintf(`SELECT count(*) FROM %[1]s.%[2]s`, bq.Namespace, tableName)
+	it, err := bq.Db.Query(sqlStatement).Read(bq.BQContext)
```
Comment (sivashanmukh) [inline warehouse/bigquery/bigquery.go:478]: Handle this error

File: warehouse/upload.go [hunk 6]
```diff
@@ -1305,30 +1405,12 @@ func (job *UploadJobT) GetLoadFileLocations(tableName string) (locations []strin
...
-		job.warehouse.Source.ID,
-		job.warehouse.Destination.ID,
-		tableName,
-		job.upload.StartLoadFileID,
-		job.upload.EndLoadFileID,
-	)
-	err = dbHandle.QueryRow(sqlStatement).Scan(&location)
-	if err != nil && err != sql.ErrNoRows {
-		pkgLogger.Errorf(`[WH] Error querying for sample load file location: %v`, err)
+	var locations []string
```
Comment (sivashanmukh) [inline warehouse/upload.go:1408]: Merge this with the next line?

File: warehouse/bigquery/bigquery.go (hunk 5 above)
Comment (sivashanmukh) [inline warehouse/bigquery/bigquery.go:478]: Handle this error

**Review by sivashanmukh — CHANGES_REQUESTED** [review]

**Review by sivashanmukh — APPROVED** [review]
LGTM

**Review by SumanthPuram — APPROVED** [review]

**Pull request merged**